### PR TITLE
gh-79009: sqlite3.iterdump now correctly handles tables with autoincrement

### LINKED ISSUE #79009: sqlite.iterdump does not work for (most) databases with autoincrement
BPO | [34828](https://bugs.python.org/issue34828)
--- | :---
Nosy | @berkerpeksag, @itssme, @erlend-aasland
PRs | <li>python/cpython#9621</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2018-09-28.06:51:55.269>
labels = ['3.7', '3.8', 'type-bug', 'library']
title = 'sqlite.iterdump does not work for (most) databases with autoincrement'
updated_at = <Date 2021-07-15.21:18:56.228>
user = 'https://github.com/itssme'
```

bugs.python.org fields:
```python
activity = <Date 2021-07-15.21:18:56.228>
actor = 'erlendaasland'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2018-09-28.06:51:55.269>
creator = 'itssme'
dependencies = []
files = []
hgrepos = []
issue_num = 34828
keywords = ['patch']
message_count = 3.0
messages = ['326610', '326611', '326657']
nosy_count = 3.0
nosy_names = ['berker.peksag', 'itssme', 'erlendaasland']
pr_nums = ['9621']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue34828'
versions = ['Python 3.6', 'Python 3.7', 'Python 3.8']
```

</p></details>

### PR DESCRIPTION
**Pull request by itssme**
The iterdump command in Lib/sqlite3/dump.py assumed that the table "sqlite_sequence" already exists. When this was not the case and exception was thrown.

Changes:
* Sqlite3 iterdump now deletes and recreates the "sqlite_sequence" table at the end of the transaction (Just like the sqlite3 command ".dump" does)
* Added two new test for the iterdump command

Resolves gh-79009

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by j0hax** [comment]
Hello, it seems that I stumbled upon this bug today with Python 3.6.13. Are there any news as to a fix?

**Comment by erlend-aasland** [comment]
> Hello, it seems that I stumbled upon this bug today with Python 3.6.13. Are there any news as to a fix?

3.6 is in security fix only (see [PEP 494](https://www.python.org/dev/peps/pep-0494/)). If this gets merged, it will be in 3.11, 3.10 and _maybe_ 3.9.

**Comment by itssme** [comment]
I saw that recently the unit tests of sqlite3 `Lib/sqlite3/test/dump.py` were changed, https://github.com/python/cpython/commit/849e339a925fb398be31e566c31b99ce6cdb6006#diff-641ffce15acf9f552596b6192c714f53590980dba32625057918b53a9676f9ae
I will update my PR soon.

**Comment by erlend-aasland** [comment]
> I saw that recently the unit tests of sqlite3 `Lib/sqlite3/test/dump.py` were changed, [849e339#diff-641ffce15acf9f552596b6192c714f53590980dba32625057918b53a9676f9ae](https://github.com/python/cpython/commit/849e339a925fb398be31e566c31b99ce6cdb6006#diff-641ffce15acf9f552596b6192c714f53590980dba32625057918b53a9676f9ae)
> I will update my PR soon.

Thanks. Please also sign the CLA before we can do a review.

https://devguide.python.org/pullrequest/

File: Misc/NEWS.d/next/Library/2018-09-28-22-18-03.bpo-34828.5Zyi_S.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Sqlite3.iterdump now correctly deletes the "sqlite_sequence" table and
+inserts the rows from the dumped db. Just like the command ".dump" in
+sqlite3 does.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2018-09-28-22-18-03.bpo-34828.5Zyi_S.rst]: I'd shorten this to something more accessible. For example:

    :meth:`~sqlite3.Connection.iterdump` now handles databases that use ``AUTOINCREMENT`` in one or more tables.

See the [dev guide](https://devguide.python.org/documenting/#additional-markup-constructs) for how to add reST markups.

File: Lib/sqlite3/test/dump.py [hunk 2]
```diff
@@ -1,7 +1,8 @@
 # Author: Paul Kippes <kippesp@gmail.com>
 
-import unittest
 import sqlite3 as sqlite
+import unittest
+
```
Comment (erlend-aasland) [inline Lib/sqlite3/test/dump.py]: Please remove any unrelated style changes from the PR. Ditto for added newlines (line 135, 144 and 149).

File: Lib/sqlite3/test/dump.py [hunk 3]
```diff
@@ -49,6 +50,67 @@ def test_table_dump(self):
...
+        i = self.cx.iterdump()
+        actual_sqls = [s for s in i]
+        # the NULL value should now be automatically be set to 1
+        expected_sqls[3] = expected_sqls[3].replace("NULL", "1")
+        expected_sqls = ['BEGIN TRANSACTION;'] + expected_sqls + \
+                        ['DELETE FROM "sqlite_sequence";'] + \
+                        ["""INSERT INTO "sqlite_sequence" VALUES('tags',1);"""] + \
+                        ['COMMIT;']
+        [self.assertEqual(expected_sqls[i], actual_sqls[i])
+         for i in range(len(expected_sqls))]
```
Comment (erlend-aasland) [inline Lib/sqlite3/test/dump.py]: Please format the new code according to [PEP 8](https://www.python.org/dev/peps/pep-0008/). Even if large parts of these files _do not_ follow PEP 8, we strive to follow PEP 8 (and PEP 7) when adding new code, or modifying existing code :)

File: Misc/NEWS.d/next/Library/2018-09-28-22-18-03.bpo-34828.5Zyi_S.rst (hunk 1 above)
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2018-09-28-22-18-03.bpo-34828.5Zyi_S.rst]: You can also add _"Patch by Joel Klimont"_ at the end of the NEWS entry.

**Review by erlend-aasland — CHANGES_REQUESTED** [review]

**Comment by itssme** [comment]
I will make the requested changes as soon as I am home from my holidays :+1:. Right now the internet is so slow, I cannot clone the project

**Comment by itssme** [comment]
I will make the requested changes as soon as I am home from my holidays :+1:. Right now the internet is so slow, I cannot clone the project

File: Lib/sqlite3/dump.py [hunk 4]
```diff
@@ -67,4 +72,8 @@ def _iterdump(connection):
     for name, type, sql in schema_res.fetchall():
         yield('{0};'.format(sql))
 
+    # Yield statements concerning the sqlite_sequence table at the end of the transaction: (bpo-34828)
```
Comment (erlend-aasland) [inline Lib/sqlite3/dump.py]: Is there a reason to yield these at the very end? Why not loop and yield after line 33?

File: Lib/sqlite3/test/test_dump.py [hunk 5]
```diff
@@ -49,6 +49,76 @@ def test_table_dump(self):
...
+            """CREATE TABLE "tags" ( """
+            """id integer primary key autoincrement,"""
+            """tag varchar(256),"""
+            """post int references posts);""",
+            """INSERT INTO "tags" VALUES(NULL,'test',0);""",
+            """CREATE TABLE "tags2" ( """
+            """id integer primary key autoincrement,"""
+            """tag varchar(256),"""
+            """post int references posts);"""
+        ]
```
Comment (erlend-aasland) [inline Lib/sqlite3/test/test_dump.py]: Can you wrap each statement on a line for itself. For one-liner string literals, please just use a pair of double-quotes for the string literal.

Also, would it be easier to just define this as a string with newlines, instead of a list of strings? For example:
```python3
        expected_sqls = """
            create table "posts" (it int primary key);
            insert into "posts" values(0);
            create table "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
            insert into "tags" values(NULL, 'test', 0);
            create table "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
        """

        self.cu.executescript(expected_sqls)

        # etc...
```

Ditto for `test_dump_autoincrement_create_new_db`.

File: Lib/sqlite3/test/test_dump.py [hunk 6]
```diff
@@ -49,6 +49,76 @@ def test_table_dump(self):
...
+                                VALUES(NULL,'test{0}',0);""".format(i))
+        for i in range(1, 5):
+            old_db.append("""INSERT INTO "tags2"
+                                VALUES(NULL,'test{0}',0);""".format(i))
+        old_db.append("COMMIT;")
+
+        for sql_statement in old_db:
+            self.cu.execute(sql_statement)
+
+        cx2 = sqlite.connect(":memory:")
```
Comment (erlend-aasland) [inline Lib/sqlite3/test/test_dump.py]: Add `from .test_dbapi import managed_connect` to the top of `test_dump.py`, and use it here. See examples in `Lib/sqlite3/test/test_dbapi.py`.

```python3
        with managed_connect(":memory:", in_mem=True) as cx:
            ...
```

You can then drop the explicit closing of `cu2` and `cx2`

File: Lib/sqlite3/test/test_dump.py [hunk 7]
```diff
@@ -49,6 +49,76 @@ def test_table_dump(self):
...
+
+        # the NULL value should now be automatically be set to 1
+        expected_sqls[3] = expected_sqls[3].replace("NULL", "1")
+        expected_sqls = ['BEGIN TRANSACTION;'] + expected_sqls + \
+                        ['DELETE FROM "sqlite_sequence";'] + \
+                        ["""INSERT INTO "sqlite_sequence" VALUES('tags',1);"""] + \
+                        ['COMMIT;']
+
+        for i in range(len(expected_sqls)):
+            self.assertEqual(expected_sqls[i], actual_sqls[i])
```
Comment (erlend-aasland) [inline Lib/sqlite3/test/test_dump.py]:
```suggestion
        self.assertEqual(expected_sqls, actual_sqls)
```

File: Lib/sqlite3/test/test_dump.py [hunk 8]
```diff
@@ -49,6 +49,76 @@ def test_table_dump(self):
...
+        for i in range(1, 5):
+            old_db.append("""INSERT INTO "tags2"
+                                VALUES(NULL,'test{0}',0);""".format(i))
+        old_db.append("COMMIT;")
+
+        for sql_statement in old_db:
+            self.cu.execute(sql_statement)
+
+        cx2 = sqlite.connect(":memory:")
+        query = "".join(line for line in self.cx.iterdump())
```
Comment (erlend-aasland) [inline Lib/sqlite3/test/test_dump.py]:
```suggestion
        query = "".join(self.cx.iterdump())
```

**Review by erlend-aasland — CHANGES_REQUESTED** [review]
Thanks for updating your PR :) I've left some comments.

**Comment by erlend-aasland** [comment]
@itssme, do you plan to follow up this PR?

**Comment by itssme** [comment]
Thanks for the reminder, yes I will update it this weekend :+1:

File: Lib/sqlite3/test/test_dump.py (hunk 5 above)
Comment (itssme) [reply Lib/sqlite3/test/test_dump.py]: has been changed in newest commit

File: Lib/sqlite3/test/test_dump.py (hunk 7 above)
Comment (itssme) [reply Lib/sqlite3/test/test_dump.py]: has been changed in newest commit

File: Lib/sqlite3/test/test_dump.py (hunk 8 above)
Comment (itssme) [reply Lib/sqlite3/test/test_dump.py]: has been changed in newest commit

File: Lib/sqlite3/test/test_dump.py (hunk 6 above)
Comment (itssme) [reply Lib/sqlite3/test/test_dump.py]: has been changed to `with memory_database() as cx2:` according to the implementation and examples in `test_dbapi.py`

File: Lib/sqlite3/dump.py (hunk 4 above)
Comment (itssme) [reply Lib/sqlite3/dump.py]: Currently the problem is that we assume, that the sqlite sequence table already exists. However, if we try to `delete from` the table to clear it and (by chance) it does not exist yet (no table with an autoincrement exists in the db) we get an error when running the sql commands the iterdump provides us with.

Example:
We have run the following commands on the database:
```sql
CREATE TABLE posts (id int primary key);
INSERT INTO posts VALUES(0);
CREATE TABLE "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
INSERT INTO "tags" VALUES(NULL,'test',0);
CREATE TABLE "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
```

Now on the sqlite3 console we run the `.dump` command we are tying to mimic in pythons iterdump:
```sql
sqlite> .dump
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;
CREATE TABLE posts (id int primary key);
INSERT INTO posts VALUES(0);
CREATE TABLE IF NOT EXISTS "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
INSERT INTO tags VALUES(1,'test',0);
CREATE TABLE IF NOT EXISTS "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
DELETE FROM sqlite_sequence;
INSERT INTO sqlite_sequence VALUES('tags',1);
COMMIT;
```

As we can see the `DELETE FROM sqlite_sequence;` and the inserts into that table, are at the very end. If we run these sql commands on a new database, everything works as expected and the we get no error about deleting data from the sqlite sequence table, as it already exists, because the tables with the autoincrement are already created in the database.

If we now try to do the same thing in python, by creating a sql connection to a :memory: db and then creating the tables, inserting the data etc..:
```python
import sqlite3
conn = sqlite3.connect(":memory:")
cur = conn.cursor()
expected_sqls = [
             "CREATE TABLE \"posts\" (id int primary key);",
             "INSERT INTO \"posts\" VALUES(0);",
             "CREATE TABLE \"tags\" (id integer primary key autoincrement, tag varchar(256), post int references posts);",
             "INSERT INTO \"tags\" VALUES(NULL,'test',0);",
             "CREATE TABLE \"tags2\" (id integer primary key autoincrement, tag varchar(256), post int references posts);"
         ]
for sql_statement in expected_sqls:
     cur.execute(sql_statement)
iterdump = conn.iterdump()
actual_sqls = [sql_statement for sql_statement in iterdump]
for sql_command in actual_sqls:
    print(sql_command)
```

We see that the dump command already wants to delete the data from the sqlite_sequence table after creating the posts table. However, in the posts table no autoincrement is used, and the sqlite_sequence table does not even exist yet.

```sql
BEGIN TRANSACTION;
CREATE TABLE "posts" (id int primary key);
INSERT INTO "posts" VALUES(0);
DELETE FROM "sqlite_sequence";
INSERT INTO "sqlite_sequence" VALUES('tags',1);
CREATE TABLE "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
INSERT INTO "tags" VALUES(1,'test',0);
CREATE TABLE "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
COMMIT;
```

If we now try to run the sql commands pythons iterdump has provided us with in a completely fresh db:
```sql
sqlite> BEGIN TRANSACTION;
CREATE TABLE "posts" (id int primary key);
INSERT INTO "posts" VALUES(0);
DELETE FROM "sqlite_sequence";
INSERT INTO "sqlite_sequence" VALUES('tags',1);
CREATE TABLE "tags" (id integer primary key autoincrement, tag varchar(256), post int references posts);
INSERT INTO "tags" VALUES(1,'test',0);
CREATE TABLE "tags2" (id integer primary key autoincrement, tag varchar(256), post int references posts);
COMMIT;
Error: no such table: sqlite_sequence
```

We get an error because the sqlite_sequence table does not exist yet. That's why its necessary for the `sqlite_sequence` related sql commands to be at the very end of the sql script that iterdump and .dump produce.

**Comment by erlend-aasland** [comment]
@itssme, I won't have time to look at this for a week or two, but it is on my list :)

File: Lib/sqlite3/dump.py (hunk 4 above)
Comment (erlend-aasland) [reply Lib/sqlite3/dump.py]: Great job, thanks for investigating!

I think we should audit the whole iterdump command against the sqlite3 cli .dump command, but that is for another PR.

File: Lib/test/test_sqlite3/test_dump.py [hunk 9]
```diff
@@ -49,6 +51,59 @@ def test_table_dump(self):
...
+        self.cu.executescript("".join(old_db))
+        query = "".join(self.cx.iterdump())
+
+        with memory_database() as cx2:
+            cx2.executescript(query)
+            cu2 = cx2.cursor()
+
+            self.assertEqual(cu2.execute("SELECT \"seq\""
+                                         "FROM \"sqlite_sequence\""
+                                         "WHERE \"name\" == \"tags\"").fetchall()[0][0], 9)
```
Comment (erlend-aasland) [inline Lib/test/test_sqlite3/test_dump.py]: This will not work if SQLite is compiled with, for example, `-DSQLITE_DQS=0`; you need to use single quotes around string literals. Ditto for the other SELECT.

**Review by erlend-aasland — COMMENTED** [review]
Thanks for doing this!

I think the tests can be simplified; the `posts` table is not really needed to test this functionality, so I suggest to drop it, in order to narrow down the tests.

In `test_dump_autoincrement_create_new_db`, one test table should be sufficient, right?

I also got some style nits; do you mind if I just apply them before merge?

File: Lib/test/test_sqlite3/test_dump.py (hunk 9 above)
Comment (itssme) [reply Lib/test/test_sqlite3/test_dump.py]: Thanks for the hint, I was not aware of that. It should be fixed in the newest version :+1:

**Comment by itssme** [comment]
I removed the `posts` table, renamed the `tags` table to `t1` and simplified it. These were the tables from the database, where I first discovered the issue.

**Comment by itssme** [comment]
> I also got some style nits; do you mind if I just apply them before merge?

Yes please, feel free to make any changes :+1:

**Comment by erlend-aasland** [comment]
Thanks! I'll review the changes this coming week.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thank you so much for the report, the fix, and your patience, @itssme! Unfortunately this took way too long to land, but at last the fix is on its way to the upcoming Python 3.12, 3.11, and 3.10 releases.

Thanks again!
